DRV-86: [java] Add client-specified query timeout

lregnier opened the pull request:
[DRV-86](https://faunadb.atlassian.net/browse/DRV-86)

lregnier commented on faunadb-scala/src/main/scala/faunadb/FaunaClient.scala:
```diff
@@ -96,7 +97,7 @@ class FaunaClient private (connection: Connection) {
     */
   def query(expr: Expr)(implicit ec: ExecutionContext): Future[Value] =
-    connection.post("", json.valueToTree(expr)).toScala.map { resp =>
+    connection.post("", json.valueToTree(expr), None.asJava).toScala.map { resp =>
```
This will be handled on [DRV-87](https://faunadb.atlassian.net/browse/DRV-87)

faunaee commented:
Will there be a test that demonstrates the timeout behaviour? Is seems like the current implementation of `Count` could be used to create a query that takes long enough to reliably trip up a very short timeout.

marrony commented on faunadb-java/src/main/java/com/faunadb/client/FaunaClient.java:
```diff
@@ -171,7 +186,45 @@ public void close() {
+   * @see Value
+   * @see com.faunadb.client.query.Language
+   */
+  public CompletableFuture<Value> query(Expr expr, Duration timeout) {
```
I think you can just throw an exception case `timeout` is null

lregnier replied:
Yeah, I've considered that... I'm not 100% sure what the best practice to handle `null` inputs is these days: throw an `IllegalArgumentException` vs wrap it within an `Optional`?

I leaned towards wrapping it within an `Optional` since we have the other method that accepts an `Optional<Duration>`, but I'm fine with changing it if you have strong opinions against it!

marrony commented on faunadb-java/src/main/java/com/faunadb/client/FaunaClient.java:
```diff
@@ -171,7 +186,45 @@ public void close() {
+   * {@link Value} instances can be converted to native types. See {@link Value} class for details.
+   *
+   * @param expr the query to be executed.
+   * @param timeout the timeout for the current query. If a query timeout has been set when creating
```
I think you can mention the timeout precision is milliseconds

lregnier commented:
@marrony & @faunaee, thank you for taking the time to look at the PR! 😃 

I've addressed your comments, and I've added an _integration_ test case for checking that the timeout header is working as expected. I had originally thought of adding a _unit_ test for doing so, but given the current testing infrastructure, it represents more work at the moment.

evbruno commented on faunadb-common/src/main/java/com/faunadb/common/Connection.java:
```diff
@@ -374,7 +395,7 @@ private FullHttpRequest newRequest(HttpMethod method, String path, JsonNode body
   }
 
-  private CompletableFuture<FullHttpResponse> performRequest(final FullHttpRequest request) {
+  private CompletableFuture<FullHttpResponse> performRequest(final FullHttpRequest request, final Optional<Duration> requestQueryTimeout) {
```
Was the timeout supposed also the being handle in the [client](https://github.com/fauna/faunadb-jvm/blob/ea5935b79851ce3457dec9841d3ce202a5ccac5d/faunadb-common/src/main/java/com/faunadb/common/Connection.java#L427 ) ? Currently, we are only setting the headers, right ?

lregnier replied:
For the scope of this ticket, we're only setting the headers. If we handle it on the client side as well, we should consider the time for request/response roundtrip as well. Maybe it should be a different setting? (Arguably, clients could just timeout the future in case they want to)

The query timeout is supposed to count only the time it takes to run the query on the _server_ side. @nathaniel-may, could you confirm this is the case? :) Thanks!

lregnier commented:
Just an FYI, I've created following ticket for adding a unit test for the `faunadb.FaunaClient`: [DRV-147](https://faunadb.atlassian.net/browse/DRV-147).
